drift/lib/src/runtime/manager/manager.dart

```diff
@@ -0,0 +1,529 @@
+  @override
+  Stream<List<D>> watch() => $state.buildSelectStatement().watch();
+
+  /// Executes this statement, like [get], but only returns one
+  /// value. If the result too many values, this method will throw. If no
+  /// row is returned, `null` will be returned instead.
+  ///
+  /// See also: [getSingle], which can be used if the query will
+  /// always evaluate to exactly one row.
+  @override
+  Future<D?> getSingleOrNull() =>
+      $state.buildSelectStatement().getSingleOrNull();
+
+  /// Creates an auto-updating stream of this statement, similar to
+  /// [watch]. However, it is assumed that the query will only
+  /// emit one result, so instead of returning a `Stream<List<D>>`, this
+  /// returns a `Stream<D?>`. If the query emits more than one row at
+  /// some point, an error will be emitted to the stream instead.
+  /// If the query emits zero rows at some point, `null` will be added
+  /// to the stream instead.
+  @override
+  Stream<D?> watchSingleOrNull() =>
+      $state.buildSelectStatement().watchSingleOrNull();
+}
+
+/// A table manager with top level function for creating, reading, updating, and deleting items
+abstract class RootTableManager<
+    DB extends GeneratedDatabase,
+    T extends Table,
+    D extends DataClass,
+    FS extends FilterComposer<DB, T>,
+    OS extends OrderingComposer<DB, T>,
+    C extends ProcessedTableManager<DB, T, D, FS, OS, C, CI, CU>,
+    CI extends Function,
+    CU extends Function> extends BaseTableManager<DB, T, D, FS, OS, C, CI, CU> {
+  /// Create a new [RootTableManager] instance
+  const RootTableManager(super.$state);
+
+  /// Select all rows from the table
+  C all() => $state._getChildManagerBuilder($state);
+
+  /// Creates a new row in the table using the given function
+  ///
+  /// By default, an exception will be thrown if another row with the same
+  /// primary key already exists. This behavior can be overridden with [mode],
+  /// for instance by using [InsertMode.replace] or [InsertMode.insertOrIgnore].
+  ///
+  /// To apply a partial or custom update in case of a conflict, you can also
+  /// use an [upsert clause](https://sqlite.org/lang_UPSERT.html) by using
+  /// [onConflict]. See [InsertStatement.insert] for more information.
+  ///
+  /// By default, the [onConflict] clause will only consider the table's primary
+  /// key. If you have additional columns with uniqueness constraints, you have
+  /// to manually add them to the clause's [DoUpdate.target].
+  ///
+  /// Returns the `rowid` of the inserted row. For tables with an auto-increment
+  /// column, the `rowid` is the generated value of that column. The returned
+  /// value can be inaccurate when [onConflict] is set and the insert behaved
+  /// like an update.
+  ///
+  /// If the table doesn't have a `rowid`, you can't rely on the return value.
+  /// Still, the future will always complete with an error if the insert fails.
+  Future<int> create(D Function(CI o) f,
```

Comment:
These `create` functions require us to return a data class, right (or at least they seem to return `D`)? That's a little problematic for columns generated on the database (like defaults or auto-incrementing keys). Is there a way to let users provide companions too?

Reply:
Also probably nevermind this one, I've seen the example from the PR description which clears this up.